File: running_app/lib/models/activity/activity_record.dart
import 'package:running_app/models/account/author.dart';
import 'package:running_app/models/account/like.dart';
import 'package:running_app/models/social/post_comment.dart';

import '../account/user.dart';

class ActivityRecord {
  final String? id;
  final String? title;
  final String? sportType;
  final String? privacy;
  final String? distance;
  final int? points;
  final int? steps;
  final int? kcal;
  final String? completedAt;
  final String? avgMovingPace;
  final Author? user;
  final int? totalLikes;
  final int? totalComments;

  ActivityRecord({
     required this.title,
     required this.id,
     required this.sportType,
     required this.privacy,
     required this.distance,
     required this.steps,
     required this.points,
     required this.kcal,
     required this.completedAt,
     required this.avgMovingPace,
     required this.user,
     required this.totalLikes,
     required this.totalComments,
  });

  ActivityRecord.fromJson(Map<String, dynamic> json)
    : id = json['id'],
      title = json['title'],
      sportType = json['sport_type'],
      privacy = json['privacy'],
      distance = json['distance'].toString(),
      steps = json['steps'],
      points = json['points'],
      kcal = json['kcal'],
      user = Author.fromJson(json['user']),
      completedAt = json['completed_at'],
      avgMovingPace = json['avg_moving_pace'],
      totalLikes = json['total_likes'],
      totalComments = json['total_comments'];

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'sport_type': sportType,
      'privacy': privacy,
      'distance': distance,
      'steps': steps,
      'points_earned': points,
      'kcal': kcal,
      'completed_at': completedAt,
      'avg_moving_pace': avgMovingPace,
      'user': user,
      'total_likes': totalLikes,
      'total_comments': totalComments,
    };
  }

  @override
  String toString() {
    return 'ActivityRecord{id: $id, sportType: $sportType, privacy: $privacy, distance: $distance, steps: $steps, kcal: $kcal, completedAt: $completedAt}';
  }
}

class DetailActivityRecord extends ActivityRecord {
  final String? duration;
  final String? description;
  final List<Like>? likes;
  final List<ActivityRecordPostComment>? comments;

  DetailActivityRecord({
    String? id,
    String? title,
    String? sportType,
    String? privacy,
    String? distance,
    int? steps,
    int? points,
    int? kcal,
    String? completedAt,
    String? avgMovingPace,
    Author? user,
    int? totalLikes,
    int? totalComments,
    required this.duration,
    required this.description,
    required this.likes,
    required this.comments,
  }) : super(
    id: id,
    title: title,
    sportType: sportType,
    privacy: privacy,
    distance: distance,
    steps: steps,
    points: points,
    kcal: kcal,
    completedAt: completedAt,
    avgMovingPace: avgMovingPace,
    user: user,
    totalLikes: totalLikes,
    totalComments: totalComments,
  );

  DetailActivityRecord.fromJson(Map<String, dynamic> json)
    : duration = json['duration'],
      description = json['description'],
      likes = (json['likes'] as List<dynamic>)
        .map((likeJson) => Like.fromJson(likeJson))
        .toList(),
      comments = (json['comments'] as List<dynamic>)
        .map((commentJson) => ActivityRecordPostComment.fromJson(commentJson))
        .toList(),
      super.fromJson(json);

  @override
  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = super.toJson();
    data['duration'] = duration;
    data['description'] = description;
    data['user'] = user?.toJson();
    return data;
  }

  @override
  String toString() {
    return 'DetailActivityRecord{${super.toString()}, duration: $duration, description: $description, user: $user}';
  }
}

class CreateActivityRecord {
  final String? privacy;
  final double? distance;
  final String? duration;
  final String? sportType;
  final String? title;
  final String? description;
  final String? userId;
  final String? completedAt;

  CreateActivityRecord({
    required this.privacy,
    required this.distance,
    required this.duration,
    required this.sportType,
    required this.title,
    required this.description,
    required this.userId,
    required this.completedAt,
  });

  Map<String, dynamic> toJson() {
    return {
      'privacy': privacy,
      'distance': distance,
      'duration': duration,
      'sport_type': sportType,
      'title': title,
      'description': description,
      'completed_at': completedAt,
      'user_id': userId,
    };
  }

  @override
  String toString() {
    return '${toJson()}';
  }
}

class UpdateActivityRecord {
  final String? title;
  final String? description;
  final String? privacy;

  UpdateActivityRecord({
    required this.title,
    required this.description,
    required this.privacy,
  });

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'privacy': privacy
    };
  }

  @override
  String toString() {
    return "${toJson()}";
  }
}

File: running-app-backend/activity/serializers/activity_record.py
from rest_framework import serializers

from account.models.activity import Activity
from activity.models import ActivityRecord
from social.models import ActivityRecordPostLike
from account.serializers import UserSerializer
from account.serializers.user_abbr import UserAbbrSerializer 
from social.serializers import ActivityRecordPostCommentSerializer
from utils.pagination import CommonPagination

class ActivityRecordSerializer(serializers.ModelSerializer):
    user = UserAbbrSerializer()
    completed_at = serializers.SerializerMethodField()
    avg_moving_pace = serializers.SerializerMethodField()
    check_user_like = serializers.SerializerMethodField()

    def get_check_user_like(self, instance):
        user_id = self.context['user'].id
        post_id = instance.id
        instance =  ActivityRecordPostLike.objects\
            .filter(user_id=user_id, post_id=post_id).first()
        return instance.id if instance else None
    
    def get_avg_moving_pace(self, instance):
        return instance.avg_moving_pace_readable()

    def get_completed_at(self, instance):
        return instance.get_readable_date('completed_at')
    
    class Meta:
        model = ActivityRecord
        fields = (
            "id", "user", "sport_type", 
            "points", "distance", 'privacy',
            "avg_moving_pace", "avg_cadence", "avg_heart_rate",
            "duration", "steps", "user",
            "kcal", "completed_at",
            "total_likes", "total_comments", "check_user_like"
        )
        extra_kwargs = {
            "id": {"read_only": True},
            "user": {"read_only": True}
        }

class DetailActivityRecordSerializer(serializers.ModelSerializer):
    # user = serializers.SerializerMethodField()
    user = UserAbbrSerializer()
    completed_at = serializers.SerializerMethodField()
    sport_type = serializers.CharField(source='get_sport_type_display')
    privacy = serializers.CharField(source='get_privacy_display')
    avg_moving_pace = serializers.SerializerMethodField()
    kcal = serializers.SerializerMethodField()
    likes = serializers.SerializerMethodField()
    comments = serializers.SerializerMethodField()
    check_user_like = serializers.SerializerMethodField()

    def get_check_user_like(self, instance):
        user_id = self.context['user'].id
        post_id = instance.id
        instance =  ActivityRecordPostLike.objects\
            .filter(user_id=user_id, post_id=post_id).first()
        return instance.id if instance else None
    
    def get_author_id(self, instance):
        return instance.user.user.id
    
    def get_likes(self, instance):
        context = self.context
        exclude = context.get('exclude', [])
        print({'exclude': exclude})
        if 'likes' not in exclude:
            queryset = instance.likes.all()
            paginator = CommonPagination(page_size=100)
            paginated_queryset = paginator.paginate_queryset(queryset, self.context['request'])
            return UserAbbrSerializer(paginated_queryset, many=True, read_only=True, context={
                'request_user_id': self.context['user'].id
            }).data
        return None
    
    def get_comments(self, instance):
        context = self.context
        exclude = context.get('exclude', [])
        if 'comments' not in exclude:
            queryset = instance.comments.all()
            paginator = CommonPagination(page_size=5, page_query_param="cmt_pg")
            paginated_queryset = paginator.paginate_queryset(queryset, self.context['request'])
            return ActivityRecordPostCommentSerializer(paginated_queryset, many=True, read_only=True).data
        return None

    def get_completed_at(self, instance):
        return instance.get_readable_date_time('completed_at')

    # def get_user(self, instance):
    #     return UserSerializer(instance.user.user).data

    def get_avg_moving_pace(self, instance):
        return instance.avg_moving_pace_readable()
    
    def get_kcal(self, instance):
        return instance.kcal()
    
    class Meta:
        model = ActivityRecord
        fields = "__all__"
        extra_kwargs = {
            "id": {"read_only": True}
        }

class CreateActivityRecordSerializer(serializers.ModelSerializer):
    user_id = serializers.UUIDField()

    def create(self, validated_data):
        user_id = validated_data.pop('user_id')
        return ActivityRecord.objects.create(user_id=user_id, **validated_data)

    class Meta:
        model = ActivityRecord
        fields = "__all__"
        extra_kwargs = {
            "id": {"read_only": True},
            "user": {"read_only": True},
        }

class UpdateActivityRecordSerializer(serializers.ModelSerializer):
    class Meta:
        model = ActivityRecord
        fields = (
            "description",
            "title",
            "privacy",
        )

